x-pack/plugins/security_solution_serverless/server/cloud_security/cloud_security_metering_task.ts

```diff
@@ -0,0 +1,144 @@
+  CLOUD_SECURITY_TASK_TYPE,
+} from './cloud_security_metering';
+import { cloudSecurityMetringTaskProperties } from './cloud_security_metering_task_config';
+import type { CloudSecurityMeteringCallbackInput, ResourceCountAggregation } from './types';
+
+export const getCloudSecurityUsageRecord = async ({
+  esClient,
+  projectId,
+  logger,
+  taskId,
+  postureType,
+}: CloudSecurityMeteringCallbackInput): Promise<UsageRecord | undefined> => {
+  try {
+    if (!postureType) {
+      logger.error('posture type is missing');
+      return;
+    }
+
+    const response = await esClient.search<unknown, ResourceCountAggregation>(
+      getAggQueryByPostureType(postureType)
+    );
+
+    if (!response.aggregations) {
+      return;
+    }
+    const resourceCount = response.aggregations.unique_resources.value;
+
+    const minTimestamp = response.aggregations
+      ? new Date(response.aggregations.min_timestamp.value_as_string).toISOString()
+      : new Date().toISOString();
+
+    const usageRecords = {
+      id: `${CLOUD_SECURITY_TASK_TYPE}:${postureType}`,
+      usage_timestamp: minTimestamp,
+      creation_timestamp: new Date().toISOString(),
+      usage: {
+        type: CLOUD_SECURITY_TASK_TYPE,
+        sub_type: postureType,
+        quantity: resourceCount,
+        period_seconds: cloudSecurityMetringTaskProperties.periodSeconds,
+      },
+      source: {
+        id: taskId,
+        instance_group_id: projectId,
+      },
+    };
+
+    logger.debug(`Fetched ${postureType} metring data`);
+
+    return usageRecords;
+  } catch (err) {
+    logger.error(`Failed to fetch ${postureType} metering data ${err}`);
+  }
+};
+
+export const getAggQueryByPostureType = (postureType: string) => {
+  return postureType === 'cspm' || postureType === 'kspm'
+    ? getFindingsByResourceAggQuery(postureType)
+    : getVulnerabilitiesByResourceAggQuery();
+};
+
+export const getFindingsByResourceAggQuery = (postureType: string) => ({
+  index: LATEST_FINDINGS_INDEX_PATTERN,
+  query: {
+    bool: {
+      must: [
+        {
+          term: {
+            'rule.benchmark.posture_type': postureType,
+          },
+        },
+        {
+          range: {
+            '@timestamp': {
+              gte: 'now-' + LATEST_FINDINGS_RETENTION_POLICY,
+            },
+          },
+        },
+      ],
+    },
+  },
+  size: 0,
+  aggs: {
+    unique_resources: {
+      cardinality: {
+        field: 'resource.id',
+        precision_threshold: AGGREGATION_PRECISION_THRESHOLD,
+      },
+    },
+    min_timestamp: {
+      min: {
+        field: '@timestamp',
+      },
+    },
+  },
+});
+
+export const getVulnerabilitiesByResourceAggQuery = () => ({
+  index: LATEST_VULNERABILITIES_INDEX_PATTERN,
+  query: {
+    bool: {
+      must: [
+        {
+          range: {
+            '@timestamp': {
+              gte: 'now-' + LATEST_VULNERABILITIES_RETENTION_POLICY, // the "look back" period should be the same as the scan interval
```

Comment:
The idea I discussed with @kfirpeled is that we would charge users for the period of time they use the product, and the product remain available as long as the findings are within the TTL range. 

For example, for vulnerability management, the cycle is 24 hours, and the TTL is 72 hours, a new resource finding will be indexed after more than 24 hours.
If we will use  the cycle period as the "lock back" param, it will result in a time range where the customer does not use the product from a billing perspective even though he does use it.

Reply:
I agree, added it to https://github.com/elastic/security-team/issues/7256

Reply:
I agree with @eyalkraft , it would make more sense to use cycle interval instead of the retention policy as they differ significantly. Good call!

Until we will have a different approach we will move on with the current one